### PR TITLE
Simultaneous Frecuency shifting optimization

### PR DESCRIPTION
**Pull request by alalons**
New center frecuency at 868,625. 
Uses one function to apply two complex rotations reusing calculations, should speed up in systems with limited resources. 
Help needed to profile the program and see if it really improves. Never done that. 
Requires new frequency step at 25Khz to get to the exact frequencies displacing 325khz up and down.
Also made decimation float to test lower sample rates hoping it will require less computational power. 
Tested down to 1.15 decimation and 960000 (1.15*800khz)samples/s on S1 and C1 mode.
Test 2.7 decimation and 2160000 sample rate (2.7*800khz) best rssi levels 
Test 3 decimation 2400000 sample rate, rssi decreases

### CODE REVIEW

File: rtl_wmbus.c [hunk 1]
```diff
@@ -574,7 +574,7 @@ static void time2_algorithm_s1(unsigned bit, unsigned rssi, struct time2_algorit
 
 static int opts_run_length_algorithm_enabled = 1;
 static int opts_time2_algorithm_enabled = 1;
-static unsigned opts_decimation_rate = 2u;
+static float opts_decimation_rate = 2;
```
Comment (xaelsouth) [inline rtl_wmbus.c:564]: = 2.f

File: rtl_wmbus.c [hunk 2]
```diff
@@ -588,7 +588,7 @@ static void print_usage(const char *program_name)
     fprintf(stdout, "\t-t 0 to disable time2 algorithm\n");
     fprintf(stdout, "\t-d 2 set decimation rate to 2 (defaults to 2 if omitted)\n");
     fprintf(stdout, "\t-v show used algorithm in the output\n");
-    fprintf(stdout, "\t-s receive S1 and T1/C1 datagrams simultaneously. rtl_sdr _MUST_ be set to 868.7MHz (-f 868.7M)\n");
+    fprintf(stdout, "\t-s receive S1 and T1/C1 datagrams simultaneously. rtl_sdt _MUST_ be set to 868.7MHz\n");
```
Comment (xaelsouth) [inline rtl_wmbus.c:578]: already in master

File: rtl_wmbus.c [hunk 3]
```diff
@@ -623,7 +623,7 @@ static void process_options(int argc, char *argv[])
             }
             break;
         case 'd':
-            opts_decimation_rate = strtoul(optarg, NULL, 10);
+            opts_decimation_rate = strtof(optarg, NULL);
```
Comment (xaelsouth) [inline rtl_wmbus.c:613]: seems to be ok

File: rtl_wmbus.c [hunk 4]
```diff
@@ -640,7 +640,7 @@ static void process_options(int argc, char *argv[])
 
 static float complex *LUT_FREQUENCY_TRANSLATION_PLUS = NULL;
 static float complex *LUT_FREQUENCY_TRANSLATION_MINUS = NULL;
-#define FREQ_STEP_KHZ (50)
+#define FREQ_STEP_KHZ (25) //to acomodate symetric filter we need more resolution or maybe use floats
```
Comment (xaelsouth) [inline rtl_wmbus.c:630]: seems to be ok

File: rtl_wmbus.c [hunk 5]
```diff
@@ -712,15 +716,52 @@ static void shift_freq_minus400(float *i, float *q, int fs_kHz)
...
+	ix=*iplus * x;
+	qx=*qplus * x;
+	iz=*iplus * z;
+	qz=*qplus * z;
+	//simetric negative shift 
+	*iminus=ix+qz; 
+	*qminus=qx-iz;
+	//simetric positive shift 
+	*iplus=ix-qz;
+	*qplus=qx+iz;
```
Comment (xaelsouth) [inline rtl_wmbus.c:737]: Did'not really check this. But something like this is usually more readable

 float complex freq_shift, s;

 freq_shift = LUT_FREQUENCY_TRANSLATION_PLUS[n];
 s = (*iplus) + (*qplus) * _Complex_I;
 s = s * freq_shift;
 *iplus = crealf(s);
 *qplus = cimagef(s);

 freq_shift = conjf(freq_shift);
 s = (*iminus) + (*qminus) * _Complex_I;
 s = s * freq_shift;
 *iminus = crealf(s);
 *qminus = cimagef(s);

But it's basically the same as before: two shifts in one function vs. two shifts splitted in two functions.

File: rtl_wmbus.c [hunk 6]
```diff
@@ -766,12 +807,10 @@ int main(int argc, char *argv[])
 
         for (size_t k = 0; k < sizeof(samples)/sizeof(samples[0]); k += 2)   // +2 : i and q interleaved
         {
-            const float i_unfilt = ((float)samples[k]     - 127.5f);
-            const float q_unfilt = ((float)samples[k + 1] - 127.5f);
-
-            // rtl_sdr -f 868.35M -s 2400000 - 2>/dev/null | build/rtl_wmbus -d 3
-            //shift_freq(&i_unfilt, &q_unfilt, 600, 2400);
+            const float i_unfilt = ((float)samples[k]     - 127.5);
+            const float q_unfilt = ((float)samples[k + 1] - 127.5);
```
Comment (xaelsouth) [inline rtl_wmbus.c:799]: would be double operation. samples[k] - 127.5f should be left here.

File: rtl_wmbus.c [hunk 7]
```diff
@@ -780,9 +819,7 @@ int main(int argc, char *argv[])
 
             if (opts_s1_t1_c1_simultaneously)
             {
-                shift_freq_250(&i_t1_c1_unfilt, &q_t1_c1_unfilt, fs_kHz);
-
-                shift_freq_minus400(&i_s1_unfilt, &q_s1_unfilt, fs_kHz);
+            shift_freq_plus_minus325(&i_t1_c1_unfilt, &q_t1_c1_unfilt,&i_s1_unfilt, &q_s1_unfilt, fs_kHz);
```
Comment (xaelsouth) [inline rtl_wmbus.c:810]: As I explained above, two frequency translations were put into one function. It's an open question if this is a speed up, but surely less readable.

File: rtl_wmbus.c [hunk 8]
```diff
@@ -823,7 +860,7 @@ int main(int argc, char *argv[])
 
             ++decimation_rate_index;
             if (decimation_rate_index < opts_decimation_rate) continue;
-            decimation_rate_index = 0;
+            decimation_rate_index -= opts_decimation_rate;
```
Comment (xaelsouth) [inline rtl_wmbus.c:846]: I would not do that without a well-founded reason.

**Review by xaelsouth — COMMENTED** [review]
The patch does not really solve the execution speed issue. I can't accept the patch without any tests regarding that. The second reason is that I don't know how modifications will impact receiver quality.

File: rtl_wmbus.c (hunk 6 above)
Comment (alalons) [reply rtl_wmbus.c:799]: Had an outdated version without this

File: rtl_wmbus.c (hunk 7 above)
Comment (alalons) [reply rtl_wmbus.c:810]: Verified now with profiler gprof, speeds up but it is not the main bottleneck

File: rtl_wmbus.c (hunk 5 above)
Comment (alalons) [reply rtl_wmbus.c:737]: Learnt to use gprof.
Verified with profiler, using s1_c1_samples.bin spends 2.14seconds to apply both frequency translations.
I'm saving costly float multiplications there.
Original spends with same sample file 4.84 seconds
The proposed readable using complex type only optimizes the number of function calls and runs in 3.95 seconds

The real bottleneck are FIR and IIR filters. 

optimiced gprof
index % time    self  children               called                   name
                2.14    0.00 111542272/111542272     shift_freq_plus_minus325 [11]
proposed readable using complex type
index % time    self  children               called                   name
              3.95    0.00 111542272/111542272     shift_freq_plus_minus325 [11]

original gprof
index % time    self  children               called                   name
                        2.47    0.00    111542272/111542272     shift_freq_minus400 [12]
                        2.37    0.00    111542272/111542272     shift_freq_250 [13]

**Comment by alalons** [comment]
Closing as there were changes in master not updated in my branch. Created new pull request with only secure mathematical optimizations

**Pull request closed without merging**